### PR TITLE
Sparse Li-chao tree

### PR DESCRIPTION
**Pull request by plasmatic1**
bigger range of x-coords

tested on ceoi17p4

### CODE REVIEW

File: li_chao_tree_sparse.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        }
+        ll mid = (l + r) / 2;
+        bool lDom = ln.eval(l) < n->val.eval(l), mDom = ln.eval(mid) < n->val.eval(mid);
+        if (mDom) swap(ln, n->val);
+        if (l == r) return; // base case
+        if (lDom != mDom) _insert(ln, n->l, l, mid);
+        else _insert(ln, n->r, mid + 1, r);
+    }
+    void insert(line ln) { _insert(ln, rt, MINX, MAXX); }
+    ll _query(ll x, Node *n, ll l, ll r) {
```
Comment (wesley-a-leung) [inline li_chao_tree_sparse.cpp:29]: This should probably be changed to type T of line. Type can be inferred using `decltype`. Same with line 31, and 38.

File: li_chao_tree_sparse.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    T m, b;
+    Line(T m0, T b0) : m(m0), b(b0) {}
+    T eval(T x) { return x * m + b; }
+};
+using line = Line<ll>; // Define!!!
+const ll MINX = -1e6, MAXX = 1e6; // Define!!!
+struct LiChao { // Just *-1 if you want to do max lol
+    struct Node {
+        Node *l = nullptr, *r = nullptr;
+        line val{0, LLINF};
```
Comment (wesley-a-leung) [inline li_chao_tree_sparse.cpp:11]: `LLINF` is not currently defined anywhere. Maybe it should be a constructor argument for `LiChao`?

**Review by wesley-a-leung — CHANGES_REQUESTED** [review]

File: li_chao_tree_sparse.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        Node(Node *l0, Node *r0, line v0) : l(l0), r(r0), val(v0) {}
+    };
+    Node *rt = nullptr;
+    void _insert(line ln, Node *&n, ll l, ll r) {
+        if (l > r) return;
+        if (!n) {
+            n = new Node(nullptr, nullptr, ln);
```
Comment (crackersamdjam) [inline li_chao_tree_sparse.cpp:18]: Please use unique_ptr

**Review by crackersamdjam — CHANGES_REQUESTED** [review]
For better performance, could you store all the nodes in a vector and replace the pointers with ints representing the indices of nodes in the vector?

File: li_chao_tree_sparse.cpp (hunk 3 above)
Comment (wesley-a-leung) [reply li_chao_tree_sparse.cpp:18]: If a vector of nodes is used, then unique_ptr isn't necessary.

Comment (crackersamdjam) [reply li_chao_tree_sparse.cpp:18]: Yes, I should have kept the comments separate :monkey:

**Comment by crackersamdjam** [comment]
It’s time to close this :wesleythink:

**Pull request closed without merging**